text/0000-volatile-copy-and-set.md

```diff
@@ -0,0 +1,232 @@
+`ptr::copy_nonoverlapping_volatile` and `ptr::write_bytes_volatile`,
+respectively.
+
+# Motivation
+[motivation]: #motivation
+
+`ptr::read_volatile` and `ptr::write_volatile` were stabilized in RFC
+[1467](https://github.com/rust-lang/rfcs/pull/1467).  The stated motivation
+at the time was that this allowed "volatile access to memory-mapped I/O
+in stable code", something that was only previously possible using unstable
+intrinsics or "by abusing a bug in the `load` and `store` functions on
+atomic types which gives them volatile semantics
+([rust-lang/rust#30962](https://github.com/rust-lang/rust/pull/30962))."
+
+At the time, the decision was made not to also provide stable
+interfaces for the `volatile_copy_memory` or `volatile_set_memory`
+intrinsics, as they were "not used often" nor provided in C.
+However, when writing low-level code, it is sometimes also useful
+to be able to execute volatile copy and set operations.
+
+For example, when booting x86_64 "application processor" (AP) logical
+processors, code copies a sequence of instructions that for the AP to
+execute into a page in low physical memory, and then sends a startup
+inter-processor interrupt (SIPI) to the AP's local interrupt
+controller: the target interrupt vector number given in the SIPI is
+multiplied by the page size to determine the physical memory address
+where the AP should start executing.  So a SIPI sent to vector 7 of
+an AP causes that processor to begin executing instructions at
+physical memory address 0x7000.
+
+That is:
+
+```
+extern "C" {
+    fn copy_proto_page_to_phys_mem(src: usize, phys: u64);
+    fn send_init_ipi(cpu: u32);
+    fn send_sipi(cpu: u32, vector: u8);
+
+    static INIT_CODE: *const u8;
+    static INIT_CODE_LEN: usize;
+}
+
+// A contrived type for illustration; not actually useful.
+pub struct SIPIPage {
+    // Note that `bytes` is not visible outside of `SIPIPage`.
+    bytes: [u8; 4096],
+}
+
+impl SIPIPage {
+    // Note that the _only_ operation on the `bytes` field
+    // of `SIPIPage` is in `new`.  The compiler could, in
+    // theory, elide the `copy`.
+    pub fn new() -> SIPIPage {
+        let mut bytes = [0; 4096];
+        unsafe {
+            core::ptr::copy(INIT_CODE, bytes.as_mut_ptr(), INIT_CODE_LEN);
+        }
+        SIPIPage { bytes }
+    }
+}
+
+fn main() {
+    let proto_sipi_page = SIPIPage::new();
+    let some_core = 2;
+    unsafe {
+        copy_proto_page_to_phys_mem(&proto_sipi_page as *const _ as usize, 0x7000);
+        send_init_ipi(some_core);
+        send_sipi(some_core, 7);
+    }
+}
+```
+
+Obviously this is an unlikely way of initializing the SIPI page and
+a real kernel would not do it this way.
+
+Hoever, this code snippet is specifically constructed such that the
+sequence of sending IPIs makes no reference to `proto_sipi_page` and
+since the `bytes` field is not visible outside of `new`, this
+illustrates a situation in which the compiler _could_ theoretically
+elect to elide the copy.
```

Comment:
It seems like the problem isn't with the memcpy, but with sending the IPI. The IPI needs to be a release operation (from the hardware's view and the compiler's). I'm not familiar with x86 and you haven't given the implementation of `send_sipi` so I don't know if it is already. If it's not, you might just need some sort of barrier between the mempcy and sending the IPI.

Reply:
Indeed, the motivating example is flawed.

In this case, `copy_proto_page_to_phys_mem` is apparently an extern function that's called with a pointer to `proto_sipi_page`, so the compiler has no choice but to ensure the contents of `proto_sipi_page` were actually written before the call.  The fact that the pointer is first casted to `usize` doesn't change that – since it's first casted to a raw pointer, Stacked Borrows would still allow `copy_proto_page_to_phys_mem` to access it.  So the example as written is 100% well-defined even without a volatile copy.

On the other hand, if this were done in a manner incompatible with Stacked Borrows, such as (I think?) by transmuting the reference to `usize`, then making the copy volatile *would not help*. 

For one thing, the example code does not copy directly from `INIT_CODE` to `proto_sipi_page`.  It copies to `bytes`, a local variable in the `new` function; that variable is then moved into a `SIPIPage`, which is a memcpy, and then moved to the caller, which is another memcpy.  If you switched to a volatile copy, you'd be ensuring the data was actually written to some stack allocation representing `bytes`... but not that `bytes` was actually copied to `proto_sipi_page`!

This distinction not just theoretical but, unfortunately, real.  If you compile the example in debug mode, the assembly emitted for `SIPIPage::new` contains a call to `core::intrinsics::copy` followed by two compiler-inserted calls to `memcpy`.  Ideally LLVM's optimizer would be able to elide both of those `memcpy`s, but if you compile in release mode, it actually only elides one of them.  There is still an intermediate copy between `bytes` and `proto_sipi_page`.

You could fix the example in practice by directly copying from `INIT_CODE` to `proto_sipi_page`.  But that wouldn't fix the UB in theory (that would result if you violated Stacked Borrows; as I said, the example as written is fine).  *If* the optimizer for some reason thought that `proto_sipi_page` was unused, it would be free to, e.g., reuse the stack slot for other data, even if you had previously made volatile writes into it.  

More generally, I'm not sure there's any way to redeem the example to actually require volatile memcpy.  If the Rust code mapped the physical memory and copied the data directly to that mapping, then you'd finally be at a point where release operations matter (as mentioned by @parched).  But those are easily obtained.  If the implementation `send_sipi` is an extern function, that would be a compiler release operation, as long as you didn't hold a borrow of the data across the call (which, again, would be UB even with volatile).  If it's based on `asm!`, you would have to mark the assembly sequence with the `memory` clobber (to indicate that, in the words of the GCC manual, it "performs memory reads or writes to items other than those listed in the input and output operands"), which is also a compiler release operation.

I'm happy to see volatile memcpy stabilized, but it seems more useful for something like IPC.

Reply:
This is why I dislike contrived code examples in places like this. :-)

What do you mean about utility for IPC, precisely? As in shared-memory message passing or something like that?

Reply:
Yes.  Caveat: Based on [this recent thread](https://github.com/rust-lang/unsafe-code-guidelines/issues/152), there was no real consensus on whether or not using `volatile` for shared memory in Rust is recommended - or even legal.  But it is what C++ people recommend.  *If* volatile is suited for shared memory, then `volatile_copy` would be an ideal way to safely copy large buffers into or out of shared memory.